### PR TITLE
Add FrameSequence object with lazy viewer state interpolation

### PR DESCRIPTION
**Pull request by tlambert03**
~Not done here, but curious how you feel about this pattern.~

In looking at the animation slider, I noticed that the `animationSliderWidget` itself had a lot of control over the creation of interpolated frames.  and that whenever anything about the full movie changes, the slider widget triggers interpolation of the full movie:
```python
        for state in self.animation._state_generator():
            self.interpol_states.append(state)
```

This PR adds a new object (`FrameSequence(Sequence[ViewerState])`).  It acts as an (immutable) sequence of `ViewerStates`.  You can index into it (`frame_sequence[n]`) to get the _nth_ interpolated `ViewerState` of the final movie... and it will only calculate the state that is being requested (and it will cache it until the underlying key_frame list changes)  You can get the length with `len(frame_sequence)`...
This means that the animation slider can be a plain `QSlider` that just sets the animation state to the corresponding index of the `FrameSequence`
 
~probably a couple lingering bugs to work out.  but wanted to get your thoughts first:~

### CODE REVIEW

File: napari_animation/animation.py [hunk 1]
```diff
@@ -2,17 +2,137 @@
...
+            self.events.n_frames(value=len(self))
+            return
+
+        f = 0
+        for kf0, kf1 in pairwise(self.key_frames):
+            for s in range(kf1.steps):
+                fraction = s / kf1.steps
+                self._keyframe_map[f] = (kf0, kf1, fraction)
+                f += 1
+        self._keyframe_map[f] = (kf1, kf1, 0)
```
Comment (alisterburt) [inline napari_animation/animation.py:55]: this and the corresponding `__getitem__` are delightful 🙂

File: napari_animation/key_frame.py [hunk 2]
```diff
@@ -43,6 +43,22 @@ def from_viewer(cls, viewer: Viewer):
...
+        viewer.camera.update(self.camera)
+        viewer.dims.update(self.dims)
+
+        for layer_name, layer_state in self.layers.items():
+            layer = viewer.layers[layer_name]
+            for key, value in layer_state.items():
+                original_value = getattr(layer, key)
+                # Only set if value differs to avoid expensive redraws
+                if not np.array_equal(original_value, value):
+                    setattr(layer, key, value)
```
Comment (alisterburt) [inline napari_animation/key_frame.py:56]: having this on the keyframe directly makes a lot of sense :)

**Review by alisterburt — APPROVED** [review]
A pleasure to review! I really like the pattern, laziness is great, general positivity in the comments below

**Comment by tlambert03** [comment]
this is all set for another look

File: napari_animation/_qt/animation_widget.py [hunk 3]
```diff
@@ -49,17 +49,21 @@ def __init__(self, viewer: Viewer, parent=None):
         )
         self.frameWidget = FrameWidget(parent=self)
         self.saveButton = QPushButton("Save Animation", parent=self)
-        self.animationsliderWidget = AnimationSliderWidget(
-            self.animation, orientation=Qt.Horizontal, parent=self
+        self.animationSlider = QSlider(Qt.Horizontal, parent=self)
+        self.animationSlider.setToolTip("Scroll through animation")
+        self.animation._generator.events.n_frames.connect(
```
Comment (alisterburt) [inline napari_animation/_qt/animation_widget.py:54]: should be `self.animation._frames` I think?
```python
~/Programming/napari-animation/napari_animation/_qt/animation_widget.py in __init__(self=<napari_animation._qt.animation_widget.AnimationWidget object>, viewer=Viewer(axes=Axes(visible=False, labels=True, col...._qt.qt_viewer.QtViewer object at 0x176369d30>>}), parent=None)
     52         self.animationSlider = QSlider(Qt.Horizontal, parent=self)
     53         self.animationSlider.setToolTip("Scroll through animation")
---> 54         self.animation._generator.events.n_frames.connect(
        self.animation._generator.events.n_frames.connect = undefined
        global e = undefined
        self.animationSlider.setMaximum = <built-in method setMaximum of QSlider object at 0x193794550>
        global e.value = undefined
     55             lambda e: self.animationSlider.setMaximum(e.value)
     56         )

AttributeError: 'Animation' object has no attribute '_generator'```

File: napari_animation/frame_sequence.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        """Get the interpolated state at frame `key` in the animation."""
+        if key < 0:
+            key += len(self)
+        if key not in self._cache:
+            try:
+                kf0, kf1, frac = self._frame_index[key]
+            except KeyError:
+                raise IndexError(
+                    f"Frame index ({key}) out of range ({len(self)} frames)"
+                )
```
Comment (alisterburt) [inline napari_animation/frame_sequence.py:87]: getting errors here when playing with the `AnimationSlider` in the gui, will look into that

**Review by alisterburt — CHANGES_REQUESTED** [review]
Played with this and generally it's awesome. I love the `FrameSequence` object, it's really elegant!

There are a few small bugs I mentioned in the comments, I submiited a PR for one of them at tlambert-forks/napari-animation#2 but the other (`AnimationSlider` range seems to be set badly`) should be fixed before merge

Thanks for this!

File: napari_animation/key_frame.py [hunk 5]
```diff
@@ -43,6 +44,47 @@ def from_viewer(cls, viewer: Viewer):
...
+        viewer : napari.viewer.Viewer
+            A napari viewer to render screenshots from.
+        canvas_only : bool, optional
+            Whether to include only the canvas (and exclude the napari
+            gui), by default True
+
+        Returns
+        -------
+        np.ndarray
+            An RGB image of shape (h, w, 4).
```
[the remark this answers is not in the document]
Comment (tlambert03) [reply napari_animation/key_frame.py:83]: lol.  good point

File: napari_animation/frame_sequence.py (hunk 4 above)
Comment (tlambert03) [reply napari_animation/frame_sequence.py:87]: all set

**Comment by tlambert03** [comment]
ok, all set again... added a few more tests to cover some of these errors

**Review by alisterburt — CHANGES_REQUESTED** [review]
awesome - just one little error in the tests which I mentioned in the review, I think that call should be replaced with a call to `animationSlider.setValue()`?

File: napari_animation/_qt/animation_widget.py [hunk 6]
```diff
@@ -143,20 +143,9 @@ def _save_callback(self, event=None):
...
-            self.animationsliderWidget.interpol_states[new_frame]
-        )
-
-        # This gets the index of the first key frame whose frame count is above new_frame
-        n_frames = self.animationsliderWidget.cumulative_frame_count
-        active_index = (n_frames > new_frame).argmax() - 1
-
-        key_frames = self.animation.key_frames
-        with key_frames.selection.events.blocker_all():
-            key_frames.selection.active = key_frames[active_index]
```
[the remark this answers is not in the document]
Comment (tlambert03) [reply napari_animation/_qt/animation_widget.py:159]: funny... I wrote the test and then immediately proceeded to break it and push it :)

[the remark this answers is not in the document]
Comment (alisterburt) [reply napari_animation/_qt/animation_widget.py:159]: 🤣

**Pull request merged**